Add `fingerprint` token filter and `openrefine` analyzer

polyfractal opened the pull request:
Adds a `fingerprint` token filter which uses Lucene's FingerprintFilter, and an `openrefine` analyzer that combines the Fingerprint filter with lowercasing and asciifolding.

The OpenRefine algo says to use ascii-folding as the last step, and the reference implementation definitely does it last.  But that doesn't make sense to me?  Surely it would be better to fold after lowercasing but before sorting/deduping?  Otherwise `"gödel godel foo bar"` will be emitted as `"bar foo godel godel "`, which seems to defeat the point of ascii-folding and deduping.

The analyzer itself just lifts the defaults from the various components and exposes them.  I figured it made more sense to default to their defaults than re-define them.

I'm open to changing any of the names.  I opted for `openrefine` rather than a `fingerprint` analyzer, since A) it's a specific implementation and B) I don't like when we share names between analyzer/tokenizer/filter...it's confusing for people :)

/cc @markharwood 

Closes #13325

jpountz commented on docs/reference/analysis/analyzers/openrefine-analyzer.asciidoc:
```diff
@@ -0,0 +1,36 @@
+
+The `openrefine` analyzer implements a
+https://github.com/OpenRefine/OpenRefine/wiki/Clustering-In-Depth#fingerprint[fingerprinting algorithm]
+which is used bythe OpenRefine project to assist in clustering.
```
missing space between 'by' and 'the'

jpountz commented:
Please give @markharwood a chance to look at it before merging, but this looks good to me.

clintongormley commented:
I'd be ok with a `fingerprint` analyzer, as it is probably more meaningful to the user than `openrefine`.  Also, OpenRefine uses a number of techniques, one of which is this process which it describes as `Fingerprint`.

I'd also add the stop words token filter (with stop words disabled by default).

markharwood commented on core/src/test/java/org/elasticsearch/index/analysis/OpenRefineFingerprintAnalyzerTests.java:
```diff
@@ -0,0 +1,68 @@
+    public void testLimit() throws Exception {
+        Analyzer a = new OpenRefineAnalyzer(' ', 3, false);
+        assertAnalyzesTo(a, "e d c b a",
+            new String[]{});
```
Just reminding myself here - the rationale is that if combined output token length exceeds max_output_size we emit no tokens rather than some truncated form?

markharwood commented:
LGTM, thanks @polyfractal

polyfractal commented:
Np @markharwood :)  What do you think about the order of the ascii-folding?  Keep as OpenRefine implements it?

@clintongormley My only major concern is that "fingerprint" is pretty general... there are a lot of different types of fingerprints.  E.g. OpenRefine goes on to list ngram and phoenetic fingerprints, and you could probably lump hash-based fingerprints in too (md5, minhash, etc).  

But I agree "openrefine" is not overly indicative as to what it does, or which fingerprint from OR is used.  I'll switch it to `fingerprint` unless we come up with a better name in the meantime.  ++stopwords, will add those.

markharwood commented:
Order of ascii-folding I am not concerned about as long as it is deterministic. I see the tokens being used for machine use rather than human consumption. When alerting users to show near-dups I would always resort to showing examples of original doc text rather than raw link tokens.

polyfractal commented:
Changes made!  Sorry, I amended my commit and pushed before I remembered that makes reviewing more difficult

- Empty stop words.  I added it after lowercasing, before fingerprint
- Renamed to `fingerprint` analyzer
- Updated docs, fixed typos
